packages/playwright-test/src/plugins/vcsPlugin.ts

```diff
@@ -0,0 +1,130 @@
+/**
+ * Copyright (c) Microsoft Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import type { PlaywrightTestConfig, TestPlugin } from '../types';
+import { createGuid } from 'playwright-core/lib/utils';
+import { spawnAsync } from 'playwright-core/lib/utils/spawnAsync';
+
+const GIT_OPERATIONS_TIMEOUT_MS = 1500;
+
+export const vcs = (options?: VCSPluginOptions): TestPlugin => {
+  return {
+    name: 'playwright-vcs-plugin',
+
+    configure: async (config: PlaywrightTestConfig, configDir: string) => {
+      options = options || {};
+      let revision: Partial<Revision> | undefined = options.revision;
+      if (shouldRunGit(options.revision, options.mode)) {
+        revision = {
+          ...await gitStatusFromCLI(options.directory || configDir),
+          ...revision,
+        };
+      }
+      const links = linksFromEnv();
+      // Merge it all together minding the order of precedence
+      // Be sure to mutate the config passed in and not return a new copy
+      config.metadata = config.metadata || {};
+      config.metadata.generatedAt = config.metadata?.generatedAt ?? Date.now();
+      config.metadata.revision = {
+        ...config.metadata?.revision,
+        link: links.revision,
+        ...revision,
+      };
+      config.metadata.ci = {
+        ...config.metadata?.ci,
+        link: links.ci,
+        ...options.ci,
+      };
+
+      // Normalize dates
+      const timestamp = config.metadata.revision?.timestamp;
+      if (timestamp instanceof Date)
+        config.metadata.revision.timestamp = timestamp.getTime();
+
+    },
+  };
+};
+
+export type VCSPluginOptions = {
+    mode?: 'disable-cli',
+    directory?: string,
+    revision?: Partial<Revision>,
+    ci?: Partial<CI>,
+};
+
+interface Revision {
```

Comment:
This looks familiar, but it makes sense here! Also not a fan of partial.

Reply:
I flattened it here, too to see if that made sense. We'll judge in the next patchset.